### PR TITLE
s4u2proxy: only send client authdata in initial request

### PR DESCRIPTION
**Pull request by iboukris**
I think this should suffice for now, skipping the realm identification part as I hope it would be simpler later. I'll try to add a test though. (see #983).

### CODE REVIEW

**Comment by iboukris** [comment]
Added a test for the regression, and also ported it locally to before rbcd changes and it passed.

File: src/tests/s4u2proxy.c [hunk 1]
```diff
@@ -66,6 +87,12 @@ main(int argc, char **argv)
     krb5_keytab defkt;
     krb5_creds mcred, ev_cred, *new_cred;
     krb5_ticket *ev_ticket;
+    krb5_authdata **req_authdata = NULL;
+
+    if (argc == 4) {
+        req_authdata = make_request_authdata(-2, argv[3]);
```
Comment (iboukris) [inline src/tests/s4u2proxy.c]: I'm trying to change this to [ad-type ad-contents].

**Comment by greghudson** [comment]
> I think this should suffice for now, skipping the realm identification part as I hope it would be simpler later.

By realm identification part, you mean cross-realm S4U2Self?  It seems like that should just be an adjustment to the referral loop in krb5_get_self_cred_from_kdc().

**Comment by iboukris** [comment]
> By realm identification part, you mean cross-realm S4U2Self? It seems like that should just be an adjustment to the referral loop in krb5_get_self_cred_from_kdc().

No, I only looked at S4U2Proxy but I'll take a look at the self code.
I meant in rbcd when chasing referrals as impersonator to identify target proxy's realm and get a tgt to it. Since we discard the ticket it didn't seem important to send the authdata, but I'll look again.

**Comment by greghudson** [comment]
> I meant in rbcd when chasing referrals as impersonator to identify target proxy's realm and get a tgt to it. Since we discard the ticket it didn't seem important to send the authdata, but I'll look again.

Ah, I think of that as getting the target realm TGT, although it also serves to identify the realm. 
 You are correct, we don't need to (and shouldn't) include the authdata for that part of the process.

**Comment by iboukris** [comment]
Added s4u2self fixes, still untested though.

**Comment by iboukris** [comment]
I tried to add tests for the changes of not resending request authdata in cross-realm S4U referral chasing, but got into more issues related to kdb plugin handling I think. I pushed the partial test as wip commit, but I think we shouldn't wait on it as it may take long.

**Comment by iboukris** [comment]
I got the cross-realm S4U2Self test to work as expected, leaving the RBCD test as a comment for now.

**Comment by iboukris** [comment]
I figured it's our old friend 'only_pac_p()' striking again, added a wip commit that shows all is fine with the authdata handling, i can workaround it in a couple of ways but i'd like to address it, maybe we can make an assumption that if we have a PAC and krb5_db_get_authdata_info() is implemented and returned ok, then we can skip ?
Also added a commit about enterprise name canon handling that I borrowed from #983 as it would be good to get in sooner (we already document it i think, i'll look again).

**Comment by iboukris** [comment]
> maybe we can make an assumption that if we have a PAC and krb5_db_get_authdata_info() is implemented and returned ok, then we can skip ?

Actually I'm starting to think the reason for "if it is the only authorization data element" was to not possibly break it for windows client who may not understand signedpath authdata, so basically we always fine if see a PAC, but if the PAC is not the only ad then we can add and expect signedpath, except we don't add signedpath in cross-realm, so we should just trust the PAC as the patch currently does.

**Comment by greghudson** [comment]
@lhoward: we're looking at the only_pac_p() predicate added in release 1.8, where both verification and addition of ad-signedpath are skipped if the authdata contains a PAC and nothing else.  Do you recall the reasoning behind that check?

(I haven't followed exactly how this is getting in the way of these tests.  I assume that the addition of client-requested authdata causes only_pac_p() to return false, and therefore signedpath handling kicks in, but I'm not sure how that breaks the test.)

**Comment by iboukris** [comment]
> I assume that the addition of client-requested authdata causes only_pac_p() to return false, and therefore signedpath handling kicks in, but I'm not sure how that breaks the test.

Yeah, my understanding is that signedpath can't work in cross ream as it only signs with local tgt.

**Comment by lhoward** [comment]
I think it may have had something to do with Windows (or at least some release of it) getting confused with non-PAC AD types. Or simply have been because it’s redundant as the PAC has a KDC signature. It’s a long time ago, but I’ll check my email archives when I get to work. What does Heimdal do?

**Comment by iboukris** [comment]
I don't recall something like that in heimdal (pac only ad) but i'll look again, the issue at hand however is irrelevant to heimdal as we don't have rbcd yet so s4u2proxy is only local (i'd guess this was the assumption here to).

**Comment by iboukris** [comment]
Yeah, in heimdal we set ad_signedpath for s4u2proxy based either on PAC or on signedpath (the PAC verification is more explicit in heimdal though).

**Comment by iboukris** [comment]
btw, at some point maybe we should design KRB5SignedPathV2, to allow rbcd interoperability with heimdal.

**Comment by iboukris** [comment]
I think the concern about S4U2Proxy, is to make sure the kdc has issued the 2nd ticket and not made by the impersonating service (like kimpersonate).
In cross-realm RBCD we require the authorization-data of the 2nd ticket to be verified by krb5_db_get_authdata_info() in order to get the reply client principal, so it should be fine, but maybe we should make it clearer that if krb5_db_get_authdata_info() return ad_info (or success) it means the authdata were verified, then we can just check that there was a PAC in the authorization-data.

**Comment by iboukris** [comment]
I added a fixup commit the way I suggested above (and removed the kdc enterprise name canon handling commit for now, will submit it separately).

**Comment by iboukris** [comment]
I'm having second thoughts about not assuming KDC_OPT_CANONICALIZE in S4U, maybe we should leave the assumption it in the sync _for_user/_for_proxy calls but not in the stepwise layer (and krb5_get_credentials for proxy), I'll update the PR shortly.

**Comment by iboukris** [comment]
I settled with removing only the S4U2Self commit for now, the cross-realm S4U2Proxy couldn't never work unless the flag was requested so I think it's ok.

**Comment by iboukris** [comment]
> It’s a long time ago, but I’ll check my email archives when I get to work.

@lhoward ping, any chance you got something on that, thanks!

**Comment by lhoward** [comment]
> > It’s a long time ago, but I’ll check my email archives when I get to work.
> 
> @lhoward ping, any chance you got something on that, thanks!

Will follow up via email.

**Comment by lhoward** [comment]
Summary: circa Windows 2000, the Windows LSA would crash if you returned additional authorization data. I recall triggering this, returning a proposed NFSv4 ACE type from XAD. So conservatively we avoided generating `KRB5SignedPath` if the PAC is the *only* element in the authorization data. Things may have changed now so, if Windows 2000 interoperability is no longer a concern, likely you can return both.

**Comment by iboukris** [comment]
Looking at some packet captures, moderns clients are certainly fine with more authorization data elements besides the PAC, I'll test adding KRB5SignedPath to make sure, and revisit the PR, thanks!

**Comment by iboukris** [comment]
I've tested with samba and a windows 2019 as client, and AD-SIGNTICKET (512) has no impact.
BTW, i noticed this ad-type used to conflict with Windows, maybe related:
https://github.com/krb5/krb5-assignments/blob/master/ad-type

**Comment by iboukris** [comment]
Also experimented with KDB_CAP_HANDLE_ENTERPRISE, some tests needed update so it does constitute a change, but not much (I left the ldap module though).

**Comment by iboukris** [comment]
Revised as discussed, removed the cross-realm RBCD tests and the capabilities, but kept the changes in the tests as they harmless.

**Comment by greghudson** [comment]
I believe I'm done revising this, and will merge it soon barring feedback.

**Comment by iboukris** [comment]
Looks good, thanks.

**Pull request merged**